pkg/logql/functions.go

```diff
@@ -128,12 +128,85 @@ func rateLogs(selRange time.Duration, computeValues bool) func(samples []promql.
 		if !computeValues {
 			return float64(len(samples)) / selRange.Seconds()
 		}
-		var total float64
-		for _, p := range samples {
-			total += p.V
+		return extrapolatedRate(samples, selRange, true, true)
+	}
+}
+
+// extrapolatedRate is a utility function for rate/increase/delta.
+// It calculates the rate (allowing for counter resets if isCounter is true),
+// extrapolates if the first/last sample is close to the boundary, and returns
+// the result as either per-second (if isRate is true) or overall.
+func extrapolatedRate(samples []promql.Point, selRange time.Duration, isCounter, isRate bool) float64 {
```

Comment:
Could you please add a comment that this is taken from prometheus code base and where, this can be useful for everyone.

Reply:
done ，thanks